File: common/src/main/java/com/nmh/commerce/domain/Money.java
package com.nmh.commerce.domain;

import lombok.Getter;

import java.math.BigDecimal;
import java.math.RoundingMode;

@Getter
public class Money {
    public static final Money ZERO = Money.of(BigDecimal.ZERO);

    private final BigDecimal value;

    private Money(BigDecimal value) {
        this.value = value;
    }

    public static Money of(BigDecimal value) {
        return new Money(value);
    }

    public Money plus(Money other) {
        return new Money(value.add(other.value));
    }

    public Money minus(Money other) {
        return new Money(value.subtract(other.value));
    }

    public Money multiply(Money other) {
        return new Money(value.multiply(other.value));
    }

    public Money divide(Money other) {
        return new Money(value.divide(other.value, 0, RoundingMode.HALF_UP));
    }

    public boolean isLessThan(Money other) {
        return value.compareTo(other.value) < 0;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return value.compareTo(other.value) >= 0;
    }

    public Money apply(DiscountRate rate) {
        return new Money(value.multiply(rate.getValue()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Money money = (Money) o;
        return value.compareTo(money.value) == 0;
    }

    @Override
    public int hashCode() {
        return value.stripTrailingZeros().hashCode();
    }
}

File: common/src/main/kotlin/com/nmh/commerce/domain/Money.kt
package com.nmh.commerce.domain

import java.math.BigDecimal
import java.math.RoundingMode

class Money private constructor(
    val value: BigDecimal,
) {
    fun plus(other: Money): Money = Money(value.add(other.value))

    fun minus(other: Money): Money = Money(value.subtract(other.value))

    fun multiply(other: Money): Money = Money(value.multiply(other.value))

    fun divide(other: Money): Money = Money(value.divide(other.value, 0, RoundingMode.HALF_UP))

    fun isLessThan(other: Money): Boolean = value.compareTo(other.value) < 0

    fun isGreaterThanOrEqual(other: Money): Boolean = value.compareTo(other.value) >= 0

    fun apply(rate: DiscountRate): Money = Money(value.multiply(rate.value))

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val money = o as Money
        return value.compareTo(money.value) == 0
    }

    override fun hashCode(): Int = value.stripTrailingZeros().hashCode()

    companion object {
        val ZERO: Money = of(BigDecimal.ZERO)

        fun of(value: BigDecimal): Money = Money(value)
    }
}
